cdc/sorter/leveldb/leveldb.go

```diff
@@ -0,0 +1,219 @@
+	return &LevelActor{
+		id:       actor.ID(id),
+		db:       db,
+		wb:       wb,
+		iterSema: iterSema,
+		wbSize:   wbSize,
+		closedWg: wg,
+
+		metricWriteDuration: sorterWriteDurationHistogram.WithLabelValues(captureAddr, idTag),
+		metricWriteBytes:    sorterWriteBytesHistogram.WithLabelValues(captureAddr, idTag),
+	}, mb, nil
+}
+
+func (ldb *LevelActor) close(err error) {
+	log.Info("leveldb actor quit", zap.Uint64("ID", uint64(ldb.id)), zap.Error(err))
+	ldb.closedWg.Done()
+}
+
+func (ldb *LevelActor) maybeWrite(force bool) error {
+	bytes := len(ldb.wb.Dump())
+	if bytes >= ldb.wbSize || (force && bytes != 0) {
+		startTime := time.Now()
+		err := ldb.db.Write(ldb.wb, nil)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		ldb.metricWriteDuration.Observe(time.Since(startTime).Seconds())
+		ldb.metricWriteBytes.Observe(float64(bytes))
+
+		// Reset write batch or reclaim memory if it grows too large.
+		if cap(ldb.wb.Dump()) <= 2*ldb.wbSize {
+			ldb.wb.Reset()
+		} else {
+			ldb.wb = leveldb.MakeBatch(ldb.wbSize)
+		}
+	}
+	return nil
+}
+
+// Poll implements actor.Actor.
+// It handles tasks by writing kv, deleting kv and taking iterators.
+func (ldb *LevelActor) Poll(ctx context.Context, tasks []actormsg.Message) bool {
+	select {
+	case <-ctx.Done():
+		ldb.close(ctx.Err())
+		return false
+	default:
+	}
+	iterChs := make([]chan message.LimitedIterator, 0, len(tasks))
+	iterRanges := make([]*util.Range, 0, len(tasks))
+	for i := range tasks {
+		var task message.Task
+		msg := tasks[i]
+		switch msg.Tp {
+		case actormsg.TypeSorterTask:
+			task = msg.SorterTask
+		case actormsg.TypeStop:
+			ldb.close(nil)
+			return false
+		default:
+			log.Panic("unexpected message", zap.Any("message", msg))
+		}
+		events, needIter, iterCh := task.Events, task.NeedIter, task.IterCh
+
+		for k, v := range events {
+			if len(v) != 0 {
+				ldb.wb.Put([]byte(k), v)
+			} else {
+				// Delete the key if value is empty
+				ldb.wb.Delete([]byte(k))
+			}
+
+			// Do not force write, batching for efficiency.
+			if err := ldb.maybeWrite(false); err != nil {
+				log.Panic("leveldb error", zap.Error(err))
+			}
+		}
+		if needIter {
+			// Append to slice for later batch acquiring iterators.
+			iterChs = append(iterChs, iterCh)
+			iterRanges = append(iterRanges, task.Irange)
+		} else {
+			// Or close channel to notify caller that that its task is done.
+			close(iterCh)
+		}
+	}
+
+	// Force write only if there is a task requires an iterator.
+	forceWrite := len(iterChs) != 0
+	if err := ldb.maybeWrite(forceWrite); err != nil {
+		log.Panic("leveldb error", zap.Error(err))
+	}
+	// Batch acquire iterators.
+	for i := range iterChs {
+		iterCh := iterChs[i]
+		err := ldb.iterSema.Acquire(ctx, 1)
+		if err != nil {
+			if errors.Cause(err) == context.Canceled {
+				return false
+			}
+			log.Panic("leveldb error, acquire iter", zap.Error(err))
+		}
+		iter := ldb.db.NewIterator(iterRanges[i], nil)
+		iterCh <- message.LimitedIterator{
```

Comment:
Is `iterCh` a buffered channel or not?

Reply:
Yes, it's a buffered channel.
https://github.com/pingcap/ticdc/blob/b4c6b17ca6cda6ba04d6b1d1a49e9aa08c1d4d16/cdc/sorter/leveldb/message/task.go#L33-L34

Reply:
Could the buffer full cause blocking here?

Reply:
Buffer full does block, but it should not happen.